File: py-algorithm/array/intersection_of_two_arrays.py
import unittest
from typing import List
from queue import PriorityQueue
from collections import Counter

"""
    https://leetcode.com/problems/intersection-of-two-arrays/
    
    Tìm giao của 2 mảng
"""


class Solution(unittest.TestCase):

    @staticmethod
    def intersection_OK(nums1: List[int], nums2: List[int]) -> List[int]:
        nums1.sort()
        nums2.sort()
        common = set()
        i, j = 0, 0
        while i < len(nums1) and j < len(nums2):
            if nums1[i] == nums2[j]:
                common.add(nums1[i])
                i += 1
                j += 1
            elif nums1[i] < nums2[j]:
                i += 1
            else:
                j += 1

        return list(common)

    @staticmethod
    def intersection(nums1: List[int], nums2: List[int]) -> List[int]:
        return list(set(nums1) & set(nums2))


def test(self):
    self.assertEqual([1, 2], self.intersection([1, 1, 2, 3, 4], [1, 2]))


if __name__ == '__main__':
    unittest.main()

File: leetcode/src/main/java/leetcode/array/IntersectionOfTwoArrays.java
package leetcode.array;

import java.util.*;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

/**
 * https://leetcode.com/problems/intersection-of-two-arrays/description/
 *
 * Tìm giao của 2 mảng
 */
public class IntersectionOfTwoArrays {

    public int[] intersection(int[] nums1, int[] nums2) {
        Arrays.sort(nums1);
        Arrays.sort(nums2);
        Set<Integer> common = new HashSet<>();
        int i = 0, j = 0;
        while (i < nums1.length && j < nums2.length) {
            if (nums1[i] == nums2[j]) {
                common.add(nums1[i]);
                i++;
                j++;
            } else if (nums1[i] < nums2[j]) {
                i++;
            } else {
                j++;
            }
        }
        return common.stream().sorted().mapToInt(Integer::intValue).toArray();
    }

    @Test
    public void test() {
        assertArrayEquals(new int[]{2}, intersection(new int[]{1, 2, 2, 1}, new int[]{2, 2}));
        assertArrayEquals(new int[]{4, 9}, intersection(new int[]{4, 9, 5}, new int[]{9, 4, 9, 8, 4}));
    }
}
